File: pedometer.rb
# pedometer.rb
#!/usr/bin/env ruby
require 'json'
require 'date'

CONFIG_FILE = 'pedometer_config.json'
DATA_FILE = 'steps.json'

class Pedometer
  attr_reader :stride_cm, :weight_kg, :goal, :entries

  def initialize
    @stride_cm = 78.0
    @weight_kg = 70.0
    @goal = 10000
    @entries = []
    load_config
    load_entries
  end

  def load_config
    if File.exist?(CONFIG_FILE)
      cfg = JSON.parse(File.read(CONFIG_FILE))
      @stride_cm = cfg['stride_cm'] || 78.0
      @weight_kg = cfg['weight_kg'] || 70.0
      @goal = cfg['goal'] || 10000
    end
  end

  def save_config
    File.write(CONFIG_FILE, JSON.pretty_generate({
      'stride_cm' => @stride_cm,
      'weight_kg' => @weight_kg,
      'goal' => @goal
    }))
  end

  def load_entries
    if File.exist?(DATA_FILE)
      @entries = JSON.parse(File.read(DATA_FILE))
    end
  end

  def save_entries
    File.write(DATA_FILE, JSON.pretty_generate(@entries))
  end

  def get_entry(date)
    @entries.find { |e| e['date'] == date }
  end

  def add_steps(steps, date_str = nil)
    date_str ||= Date.today.to_s
    entry = get_entry(date_str)
    if entry
      entry['steps'] = steps
    else
      @entries << { 'date' => date_str, 'steps' => steps }
    end
    save_entries
    dist = steps * @stride_cm / 100000.0
    cal = steps * 0.04 * @weight_kg / 70.0
    puts "✅ Logged #{steps.to_s.reverse.gsub(/(\d{3})(?=\d)/, '\\1,').reverse} steps on #{date_str} – #{'%.2f' % dist} km, #{cal.round(0)} kcal"
  end

  def stats
    if @entries.empty?
      puts "No entries."
      return
    end
    total_steps = @entries.sum { |e| e['steps'] }
    total_dist = total_steps * @stride_cm / 100000.0
    total_cal = total_steps * 0.04 * @weight_kg / 70.0
    avg_steps = total_steps / @entries.size
    today = Date.today.to_s
    today_entry = get_entry(today)
    today_steps = today_entry ? today_entry['steps'] : 0
    pct = [100, (today_steps.to_f / @goal * 100).to_i].min
    fuel_saved = total_dist / 10.0
    co2_saved = fuel_saved * 2.31

    puts "\n🚶 Pedometer (Energy Saving)"
    puts "Today: #{today_steps.to_s.reverse.gsub(/(\d{3})(?=\d)/, '\\1,').reverse} steps"
    puts "Goal: #{@goal.to_s.reverse.gsub(/(\d{3})(?=\d)/, '\\1,').reverse} steps (#{pct}% complete)"
    puts "\n📊 Statistics"
    puts "Total steps: #{total_steps.to_s.reverse.gsub(/(\d{3})(?=\d)/, '\\1,').reverse}"
    puts "Total distance: #{'%.2f' % total_dist} km"
    puts "Total calories: #{total_cal.round(0)} kcal"
    puts "Average: #{avg_steps.to_s.reverse.gsub(/(\d{3})(?=\d)/, '\\1,').reverse} steps/day"
    puts "Energy saved: #{'%.2f' % fuel_saved} L of fuel"
    puts "CO₂ saved: #{'%.2f' % co2_saved} kg"
  end

  def list
    if @entries.empty?
      puts "No entries."
      return
    end
    sorted = @entries.sort_by { |e| e['date'] }.reverse
    puts "\n📋 Daily Log:"
    sorted.each do |e|
      steps = e['steps']
      dist = steps * @stride_cm / 100000.0
      cal = steps * 0.04 * @weight_kg / 70.0
      pct = [100, (steps.to_f / @goal * 100).to_i].min
      puts "#{e['date']} | #{steps.to_s.reverse.gsub(/(\d{3})(?=\d)/, '\\1,').reverse} steps | #{'%.2f' % dist} km | #{cal.round(0)} kcal | #{pct}%"
    end
  end

  def export(filename)
    require 'csv'
    CSV.open(filename, 'w') do |csv|
      csv << ["Date", "Steps", "Distance (km)", "Calories", "Goal %"]
      @entries.each do |e|
        steps = e['steps']
        dist = steps * @stride_cm / 100000.0
        cal = steps * 0.04 * @weight_kg / 70.0
        pct = [100, (steps.to_f / @goal * 100).to_i].min
        csv << [e['date'], steps, dist.round(2), cal.round(0), pct]
      end
    end
    puts "✅ Exported #{@entries.size} entries to #{filename}"
  end
end

if ARGV.empty?
  puts "Usage: pedometer.rb [config|add|stats|list|export]"
  exit
end

ped = Pedometer.new
cmd = ARGV.shift

case cmd
when 'config'
  while ARGV.any?
    case ARGV[0]
    when '--stride'
      ARGV.shift
      ped.stride_cm = ARGV.shift.to_f
    when '--weight'
      ARGV.shift
      ped.weight_kg = ARGV.shift.to_f
    when '--goal'
      ARGV.shift
      ped.goal = ARGV.shift.to_i
    else
      break
    end
  end
  ped.save_config
  puts "✅ Config updated: stride=#{ped.stride_cm}cm, weight=#{ped.weight_kg}kg, goal=#{ped.goal}"

when 'add'
  if ARGV.empty?
    puts "Usage: add <steps> [--date YYYY-MM-DD]"
    exit
  end
  steps = ARGV.shift.to_i
  date_str = nil
  if ARGV.include?('--date')
    idx = ARGV.index('--date')
    date_str = ARGV[idx+1] if idx
  end
  ped.add_steps(steps, date_str)

when 'stats'
  ped.stats

when 'list'
  ped.list

when 'export'
  filename = ARGV.shift || 'steps.csv'
  ped.export(filename)

else
  puts "Unknown command"
end

File: pedometer.js
// pedometer.js
#!/usr/bin/env node
const fs = require('fs');
const { program } = require('commander');

const CONFIG_FILE = 'pedometer_config.json';
const DATA_FILE = 'steps.json';

let cfg = { strideCm: 78, weightKg: 70, goal: 10000 };
let entries = [];

function loadConfig() {
    if (fs.existsSync(CONFIG_FILE)) {
        try {
            cfg = JSON.parse(fs.readFileSync(CONFIG_FILE));
        } catch (e) {}
    }
}

function saveConfig() {
    fs.writeFileSync(CONFIG_FILE, JSON.stringify(cfg, null, 2));
}

function loadEntries() {
    if (fs.existsSync(DATA_FILE)) {
        try {
            entries = JSON.parse(fs.readFileSync(DATA_FILE));
        } catch (e) {}
    }
}

function saveEntries() {
    fs.writeFileSync(DATA_FILE, JSON.stringify(entries, null, 2));
}

function getEntry(date) {
    return entries.find(e => e.date === date);
}

function formatSteps(steps) {
    return steps.toLocaleString();
}

program
    .command('config')
    .option('--stride <cm>', 'Stride length in cm', parseFloat)
    .option('--weight <kg>', 'Weight in kg', parseFloat)
    .option('--goal <n>', 'Daily step goal', parseInt)
    .action((options) => {
        if (options.stride) cfg.strideCm = options.stride;
        if (options.weight) cfg.weightKg = options.weight;
        if (options.goal) cfg.goal = options.goal;
        saveConfig();
        console.log(`✅ Config updated: stride=${cfg.strideCm}cm, weight=${cfg.weightKg}kg, goal=${cfg.goal}`);
    });

program
    .command('add <steps>')
    .option('--date <date>', 'YYYY-MM-DD')
    .action((steps, options) => {
        const s = parseInt(steps);
        const date = options.date || new Date().toISOString().slice(0,10);
        const entry = getEntry(date);
        if (entry) {
            entry.steps = s;
        } else {
            entries.push({ date, steps: s });
        }
        saveEntries();
        const dist = s * cfg.strideCm / 100000;
        const cal = s * 0.04 * cfg.weightKg / 70;
        console.log(`✅ Logged ${formatSteps(s)} steps on ${date} – ${dist.toFixed(2)} km, ${cal.toFixed(0)} kcal`);
    });

program
    .command('stats')
    .action(() => {
        if (!entries.length) {
            console.log('No entries.');
            return;
        }
        const totalSteps = entries.reduce((sum, e) => sum + e.steps, 0);
        const totalDist = totalSteps * cfg.strideCm / 100000;
        const totalCal = totalSteps * 0.04 * cfg.weightKg / 70;
        const avgSteps = Math.round(totalSteps / entries.length);
        const today = new Date().toISOString().slice(0,10);
        const todayEntry = getEntry(today);
        const todaySteps = todayEntry ? todayEntry.steps : 0;
        const pct = Math.min(100, Math.round(todaySteps / cfg.goal * 100));
        const fuelSaved = totalDist / 10;
        const co2Saved = fuelSaved * 2.31;

        console.log(`\n🚶 Pedometer (Energy Saving)`);
        console.log(`Today: ${formatSteps(todaySteps)} steps`);
        console.log(`Goal: ${formatSteps(cfg.goal)} steps (${pct}% complete)`);
        console.log(`\n📊 Statistics`);
        console.log(`Total steps: ${formatSteps(totalSteps)}`);
        console.log(`Total distance: ${totalDist.toFixed(2)} km`);
        console.log(`Total calories: ${totalCal.toFixed(0)} kcal`);
        console.log(`Average: ${formatSteps(avgSteps)} steps/day`);
        console.log(`Energy saved: ${fuelSaved.toFixed(2)} L of fuel`);
        console.log(`CO₂ saved: ${co2Saved.toFixed(2)} kg`);
    });

program
    .command('list')
    .action(() => {
        if (!entries.length) {
            console.log('No entries.');
            return;
        }
        const sorted = [...entries].sort((a, b) => b.date.localeCompare(a.date));
        console.log('\n📋 Daily Log:');
        for (const e of sorted) {
            const dist = e.steps * cfg.strideCm / 100000;
            const cal = e.steps * 0.04 * cfg.weightKg / 70;
            const pct = Math.min(100, Math.round(e.steps / cfg.goal * 100));
            console.log(`${e.date} | ${formatSteps(e.steps)} steps | ${dist.toFixed(2)} km | ${cal.toFixed(0)} kcal | ${pct}%`);
        }
    });

program
    .command('export [filename]')
    .action((filename = 'steps.csv') => {
        let csv = 'Date,Steps,Distance (km),Calories,Goal %\n';
        for (const e of entries) {
            const dist = e.steps * cfg.strideCm / 100000;
            const cal = e.steps * 0.04 * cfg.weightKg / 70;
            const pct = Math.min(100, Math.round(e.steps / cfg.goal * 100));
            csv += `${e.date},${e.steps},${dist.toFixed(2)},${cal.toFixed(0)},${pct}\n`;
        }
        fs.writeFileSync(filename, csv);
        console.log(`✅ Exported ${entries.length} entries to ${filename}`);
    });

program.parse(process.argv);
